### PR TITLE
GPU-based vectorized Specaug Version 2

### PR DESCRIPTION
**Pull request by amorari-nvidia**
# What does this PR do ?

This PR proposes a faster version of the Spectrogram Augmentation module for time and frequency masking.

### CODE REVIEW

**Review by pzelasko — DISMISSED** [review]
I visually inspected that the masks look correct on random spectrograms. I also ran a training with this variant in Nsight profiler and noticed the specaug computation mostly disappeared from the profile. It is estimated to be about 4-5x faster than the variant proposed in #9041 which would make it 8-10x faster than the original version (still available as the "legacy" setting). 

LGTM!

File: nemo/collections/asr/parts/submodules/spectr_augment.py [hunk 1]
```diff
@@ -56,7 +60,14 @@ def output_types(self):
...
     def __init__(
-        self, freq_masks=0, time_masks=0, freq_width=10, time_width=10, rng=None, mask_value=0.0,
+        self,
+        freq_masks: int = 0,
+        time_masks: int = 0,
+        freq_width: int = 10,
+        time_width: int | float = 10,
+        rng: random.Random | None = None,
+        mask_value: float = 0.0,
+        fast: bool = True,
```
Comment (titu1994) [inline nemo/collections/asr/parts/submodules/spectr_augment.py:70]: Give flag a proper name

File: nemo/collections/asr/parts/submodules/spectr_augment.py [hunk 2]
```diff
@@ -112,6 +130,82 @@ def forward(self, input_spec, length):
...
+                dtype=torch.long,
+            )
+            mask_width = torch.randint(
+                low=0, high=max(1, width), size=(batch_size, num_masks), device=input_spec.device, dtype=torch.long
+            )
+            mask_end = mask_start + mask_width
+
+        mask_end = mask_start + mask_width
+
+        indices = torch.arange(axis_length, device=input_spec.device)
```
Comment (titu1994) [inline nemo/collections/asr/parts/submodules/spectr_augment.py:196]: Can you inline document what's being done

File: nemo/collections/asr/parts/submodules/spectr_augment.py [hunk 3]
```diff
@@ -112,6 +130,82 @@ def forward(self, input_spec, length):
...
+        mask_tensor = (indices >= mask_start.unsqueeze(-1)) & (indices < mask_end.unsqueeze(-1))
+        mask_tensor = mask_tensor.any(dim=1)
+
+        mask = torch.zeros_like(input_spec, dtype=torch.bool)
+        if axis == 2:
+            mask_ranges = mask_tensor[:, None, :]
```
Comment (titu1994) [inline nemo/collections/asr/parts/submodules/spectr_augment.py:202]: Raise an error for any other value of axis

**Review by titu1994 — COMMENTED** [review]
Looks great ! Minor comments

File: nemo/collections/asr/parts/submodules/spectr_augment.py (hunk 1 above)
Comment (pzelasko) [reply nemo/collections/asr/parts/submodules/spectr_augment.py:70]: ... vectorized?

Comment (amorari-nvidia) [reply nemo/collections/asr/parts/submodules/spectr_augment.py:70]: I changed it and it gives error on a test. Doing _grep -R rng_ inside the nemo/ folder shows that the _rng_ variable is used everywhere to mean random generator, so I think it is better to leave as it is for consistency with the rest of the project.

Comment (amorari-nvidia) [reply nemo/collections/asr/parts/submodules/spectr_augment.py:70]: Never mind, I realized that I was looking at the wrong flag.

Comment (amorari-nvidia) [reply nemo/collections/asr/parts/submodules/spectr_augment.py:70]: use_vectorized_version ?

File: nemo/collections/asr/parts/submodules/spectr_augment.py (hunk 2 above)
Comment (amorari-nvidia) [reply nemo/collections/asr/parts/submodules/spectr_augment.py:196]: ok

File: nemo/collections/asr/parts/submodules/spectr_augment.py (hunk 3 above)
Comment (amorari-nvidia) [reply nemo/collections/asr/parts/submodules/spectr_augment.py:202]: ok

File: nemo/collections/asr/parts/submodules/spectr_augment.py (hunk 1 above)
Comment (amorari-nvidia) [reply nemo/collections/asr/parts/submodules/spectr_augment.py:70]: or use_vectorized_code

Comment (titu1994) [reply nemo/collections/asr/parts/submodules/spectr_augment.py]: I like use_vectorized_code

**Comment by amorari-nvidia** [comment]
After this I think that we may also have to add the flag in the audio processing module:
https://github.com/amorari-nvidia/NeMo/blob/d26eab45d54e9e00e74c68ea4b0bbbf7b48a7a50/nemo/collections/asr/modules/audio_preprocessing.py#L494

And also we may have to add a test, adding the pytest run_only_on('GPU') flag the same it is done for the numba code:
#https://github.com/amorari-nvidia/NeMo/blob/d26eab45d54e9e00e74c68ea4b0bbbf7b48a7a50/tests/collections/asr/test_asr_modules.py#L71

File: nemo/collections/asr/parts/submodules/spectr_augment.py (hunk 1 above)
Comment (amorari-nvidia) [reply nemo/collections/asr/parts/submodules/spectr_augment.py]: ok. I used that one.

File: nemo/collections/asr/parts/submodules/spectr_augment.py [hunk 4]
```diff
@@ -142,40 +142,43 @@ def _forward_fast(self, input_spec: torch.Tensor, length: torch.Tensor) -> torch
...
+                width = torch.clamp(width * length, max = axis_length).unsqueeze(1)
 
+        # Generate [0-1) random numbers and then scale the tensors.
+        # Use float32 dtype for begin/end mask markers before they are quantized to long.
+        # Using x.dtype might cause us to encounter dtypes such as bf16 or smaller which
+        # wouldn't be able to represent every frame index leading to subtle bugs.
+        mask_width = torch.rand((batch_size, num_masks), device=input_spec.device, dtype=torch.float32) * width
+        mask_width = mask_width.long()
+        mask_start = torch.rand((batch_size, num_masks), device=input_spec.device, dtype=torch.float32) * (axis_length - mask_width)
+        mask_start.long()
```
Comment (pzelasko) [inline nemo/collections/asr/parts/submodules/spectr_augment.py:180]: should we assign the result? `mask_start = mask_start.long()`

Comment (pzelasko) [reply nemo/collections/asr/parts/submodules/spectr_augment.py:180]: for reference the version that I tested and confirmed that it masks properly (ie skipping the padding regions) had the following code in this bit:

```python
if isinstance(width, float):
            width = torch.min(width * length, torch.tensor(axis_length, dtype=torch.float32)).unsqueeze(1)
        mask_width = torch.rand((batch_size, num_masks), device=input_spec.device, dtype=torch.float32) * width
        mask_width = mask_width.long()
        mask_start = torch.rand((batch_size, num_masks), device=input_spec.device, dtype=torch.float32) * (length.unsqueeze(1) - mask_width)
        mask_start = mask_start.long()
```

I see you are using axis_length rather than length for mask_start so I'm suspecting this version has the padding mask issue.

Comment (amorari-nvidia) [reply nemo/collections/asr/parts/submodules/spectr_augment.py:180]: @pzelasko you are right let me change it to length

Comment (amorari-nvidia) [reply nemo/collections/asr/parts/submodules/spectr_augment.py:180]: I fixed both

**Comment by pzelasko** [comment]
@amorari-nvidia As a last thing, could you remove the numba specaug implementation altogether in this PR? IIRC it will automatically kick in with numba 0.58 (or when numba+cuda versions are compatible for earlier versions), overriding the implementation here. CC @titu1994 pls double check if I'm right about this.

**Comment by titu1994** [comment]
You can remove it entirely, or set the default flag of numba cuda to false. I think that's sufficient for the time being. 

If you can remove it cleanly, then that's fine too

**Comment by amorari-nvidia** [comment]
I will just set the Numba version default flag to false for now

**Comment by amorari-nvidia** [comment]
The Numba code is already set to false, so this is good to go.

> I will just set the Numba version default flag to false for now

The Numba SpecAugment version is already set to false, while the vectorized version is set to true and is the default code.

**Comment by pzelasko** [comment]
great, re-triggered tests, LGTM as soon as it's green

**Comment by pablo-garay** [comment]
> great, re-triggered tests, LGTM as soon as it's green

CI passed

**Review by pzelasko — APPROVED** [review]

**Comment by pzelasko** [comment]
Congrats @amorari-nvidia on the first contribution!
